### ISSUE TITLE
memory leak in to_csv

### ISSUE THREAD

**Issue by abast** [dask/dask#1888]
The following call to DataFrame.to_csv took more than 40 GB of memory.
```
df_generator = lambda: pd.DataFrame(np.random.randint(0,100,size=(1, 5)))
ddf = [dask.delayed(df_generator)() for i in range(5000)]
ddf = dask.dataframe.from_delayed(ddf, meta = df_generator())
ddf.to_csv(os.path.join(tempfile.mkdtemp(), 'bla*.csv'))
```

The memory is allocated before a single file is written.  By interrupting the process I assume, that the memory allocation is mostly due to the function `merge `in toolz.dicttoolz. What is going on here?

Edit: I gave it some more time to run and it peaked over 130 GB memory consumption.

**Comment by mrocklin** [comment]
```python
In [1]: import numpy as np
   ...: import pandas as pd
   ...: import dask
   ...: import dask.dataframe
   ...: df_generator = lambda: pd.DataFrame(np.random.randint(0,100,size=(1, 5))
   ...: )
   ...: ddf = [dask.delayed(df_generator)() for i in range(5000)]
   ...: ddf = dask.dataframe.from_delayed(ddf, meta = df_generator())
   ...: 

In [2]: ddf.to_csv('foo2/foo.*.csv')

In [3]:                    
```

For me this peaks out at around 4GB of RAM use

**Comment by abast** [comment]
I tested it on 0.12, you probably on master?

**Comment by mrocklin** [comment]
I am, you could try 0.13, the latest release.  I don't recall any major changes to to_csv between releases, but it's worth a shot.

**Comment by abast** [comment]
I get the same behaviour on 0.13.

After interrupting, the following error stack pops up:
```
---------------------------------------------------------------------------
KeyboardInterrupt                         Traceback (most recent call last)
<ipython-input-3-0db98d7ee968> in <module>()
      2 ddf = [dask.delayed(df_generator)() for i in range(50000)]
      3 ddf = dask.dataframe.from_delayed(ddf, meta = df_generator())
----> 4 ddf.to_csv(os.path.join(tempfile.mkdtemp(), 'bla*.csv'))

[path]/Anaconda2/lib/python2.7/site-packages/dask/dataframe/core.pyc in to_csv(self, filename, **kwargs)
    936         """ See dd.to_csv docstring for more information """
    937         from .io import to_csv
--> 938         return to_csv(self, filename, **kwargs)
    939 
    940     def to_delayed(self):

[path]/Anaconda2/lib/python2.7/site-packages/dask/dataframe/io/csv.pyc in to_csv(df, filename, name_function, compression, compute, get, **kwargs)
    411     if compute:
    412         from dask import compute
--> 413         compute(*values, get=get)
    414     else:
    415         return values

[path]/Anaconda2/lib/python2.7/site-packages/dask/base.pyc in compute(*args, **kwargs)
    168         groups = {opt: [merge([v.dask for v in val]),
    169                         [v._keys() for v in val]]
--> 170                   for opt, val in groups.items()}
    171         for opt in optimizations:
    172             groups = {k: [opt(dsk, keys), keys]

[path]/Anaconda2/lib/python2.7/site-packages/dask/base.pyc in <dictcomp>((opt, val))
    168         groups = {opt: [merge([v.dask for v in val]),
    169                         [v._keys() for v in val]]
--> 170                   for opt, val in groups.items()}
    171         for opt in optimizations:
    172             groups = {k: [opt(dsk, keys), keys]

[path]/Anaconda2/lib/python2.7/site-packages/dask/delayed.pyc in dask(self)
    323     @property
    324     def dask(self):
--> 325         return merge(*self._dasks)
    326 
    327     @property

[path]/Anaconda2/lib/python2.7/site-packages/toolz/dicttoolz.pyc in merge(*dicts, **kwargs)
     36 
     37     rv = factory()
---> 38     for d in dicts:
     39         rv.update(d)
     40     return rv

KeyboardInterrupt: 
```

**Comment by abast** [comment]
I've tried it again, this time on my workstation instead of the compute server. Both use the exact same Anaconda environment (which is available through an identical nfs mounted path). However, on the workstation, it does not peak above ~4GB. Do you have any suggestions for further debugging or ideas where the problem might be?

**Comment by mrocklin** [comment]
That's confusing.  Another way to approach this is "why does this take even 4GB?  The amount of data you're generating is quite small.  You could use the dask profilers http://dask.pydata.org/en/latest/diagnostics.html#profiling to find the order in which tasks are executing and when memory spikes.

**Comment by abast** [comment]
I think, the memory peaks before anything gets submitted to the get method, in the "optimize_graph" part of the compute function:
```
if kwargs.get('optimize_graph', True):
        groups = groupby(attrgetter('_optimize'), variables)
        groups = {opt: [merge([v.dask for v in val]),
                        [v._keys() for v in val]]
                  for opt, val in groups.items()}
```
I often had the impression, that the processing of the dask graphs before the actual computation can be quite time and memory expensive and this example seems to illustrate that.

**Comment by mrocklin** [comment]
Perhaps run the cProfiler to investigate where the bottlenecks are?

    %prun ddf._optimize(ddf.dask, ddf._keys())

also ccing @jcrist

**Comment by abast** [comment]
calling `_optimize` on `ddf` or the delayed object that `read_csv(...,  compute = False)` returns is fast:

```
%time silent = ddf._optimize(ddf.dask, ddf._keys())
delayeds = ddf.to_csv(os.path.join(path, 'bla*.csv'), compute = False)
%time silent = d[0]._optimize(d[0].dask, d[0]._keys())

> Wall time: 53.7 ms
> Wall time: 839 µs
```

However, merging `delayeds `is slow:
```
%prun merge(d.dask for d in delayeds)
         
> 50008 function calls (45008 primitive calls) in 3.220 seconds

   Ordered by: internal time
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    20000    2.600    0.000    2.600    0.000 {method 'update' of 'dict' objects}
   5001/1    0.591    0.000    3.220    3.220 dicttoolz.py:19(merge)
     5001    0.011    0.000    1.309    0.000 <string>:1(<genexpr>)
     5000    0.010    0.000    1.298    0.000 delayed.py:323(dask)
     5001    0.006    0.000    0.007    0.000 dicttoolz.py:11(_get_factory)
     5001    0.001    0.000    0.001    0.000 {len}
     5001    0.001    0.000    0.001    0.000 {method 'pop' of 'dict' objects}
        1    0.001    0.001    3.220    3.220 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {isinstance}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```
The code in the mentioned section of `compute ` looks like there might be a lot to merge, so I assume this to be the bottleneck.

**Comment by abast** [comment]
Also, generally thinking about `compute`, why do we merge the graphs in case of `optimize_graph = False` instead of doing consecutive calls to the get method?

**Comment by mrocklin** [comment]
The graphs may contain repeated values.  You also might want to run multiple tasks from each graph in parallel.

**Comment by jcrist** [comment]
Should be fixed by #1894.

**Comment by abast** [comment]
Great! Compared to 0.13, the current master provides a significant speedup for me. So from my perspective, a version 0.13.1 seems desirable. Or are the changes in the delayed interface to small to justify a version 0.13.1?
